src/file_pipe_log/pipe.rs

```diff
@@ -272,31 +370,73 @@ impl<F: FileSystem> SinglePipe<F> {
 
     fn file_span(&self) -> (FileSeq, FileSeq) {
         let files = self.files.read();
-        (files.first_seq, files.active_seq)
+        (
+            files.first_seq_in_use,
+            files.first_seq + files.fds.len() as u64 - 1,
+        )
     }
 
     fn total_size(&self) -> usize {
         let files = self.files.read();
-        (files.active_seq - files.first_seq + 1) as usize * self.target_file_size
+        files.fds.len() * self.target_file_size
     }
 
     fn rotate(&self) -> Result<()> {
         self.rotate_imp(&mut self.active_file.lock())
     }
 
     fn purge_to(&self, file_seq: FileSeq) -> Result<usize> {
-        let (purged, remained) = {
+        let (
+            first_purge_seq, /* first seq for purging */
+            purged,          /* count of purged files */
+            recycled,        /* count of recycled files */
+            remained,        /* count of remained files */
+        ) = {
             let mut files = self.files.write();
-            if file_seq > files.active_seq {
+            if file_seq >= files.first_seq + files.fds.len() as u64 {
                 return Err(box_err!("Purge active or newer files"));
+            } else if file_seq <= files.first_seq_in_use {
+                return Ok(0);
             }
-            let end_offset = file_seq.saturating_sub(files.first_seq) as usize;
-            files.fds.drain(..end_offset);
-            files.first_seq = file_seq;
-            (end_offset, files.fds.len())
+
+            // If capacity == 0, `files` not support to `recycle`.
+            let logically_purged = if files.capacity == 0 {
+                0
+            } else {
+                (file_seq - files.first_seq_in_use) as usize
+            };
+            // Remove some obsolete files if capacity is exceeded.
+            let obsolete_files = (file_seq - files.first_seq) as usize;
+            // When capacity is zero, always remove logically deleted files.
+            let capacity_exceeded = files.fds.len().saturating_sub(files.capacity);
+            let purged = std::cmp::min(capacity_exceeded, obsolete_files);
+
+            // The files marked with `recycle` but with format_version V1 should also
+            // be removed.
+            files.first_seq += purged as u64;
+            let extra_purged = {
+                let mut count = 0;
+                for recycle_idx in 0..(file_seq - files.first_seq) as usize {
+                    if !files.fds[recycle_idx].context.version.has_log_signing() {
+                        count += 1;
+                    }
+                }
+                count
+            };
+            let final_purge_count = purged + extra_purged;
+            // Update metadata of files
+            files.first_seq += extra_purged as u64;
```

Comment:
The code you write is incorrect. Needs some more test coverage.
```suggestion
            // format_version V1 cannot be recycle candidate.
            // Find the newest obsolete V1 file and refresh purge count.
            for i in (purged..first_seq-files.first_seq).rev() {
              if !files.fds[i].context.version.has_log_signing() {
                purged = i + 1;
                break;
              }
            }
            files.first_seq += purged as u64;
```